content/v2.0/reference/internals/storage-engine.md

```diff
@@ -0,0 +1,106 @@
+menu:
+  v2_0_ref:
+    name: Storage engine
+    parent: InfluxDB Internals
+v2.0/tags: [storage, internals]
+---
+
+The InfluxDB storage engine ensures that
+
+- Data is safely written to disk
+- Queried data is returned complete and correct
+- Data is accurate (first) and performant (second)
+
+This document outlines the internal workings of the storage engine.
+This information is presented both as a reference and to aid those looking to maximize performance.
+
+Major topics include:
+
+* [Write Ahead Log (WAL)](#write-ahead-log-wal)
+* [Cache](#cache)
+* [Time-Structed Merge Tree (TSM)](#time-structured-merge-tree-tsm)
+* [Time Series Index (TSI)](#time-series-index-tsi)
+
+## Writing data: from API to disk
+
+The storage engine handles data from the point an API write request is received through writing data to the physical disk.
+Data is written to InfluxDB using [line protocol](/v2.0/reference/line-protocol/) sent via HTTP POST request to the `/write` endpoint.
+Batches of [points](/v2.0/reference/glossary/#point) are sent to InfluxDB, compressed, and written to a WAL for immediate durability.
+Points are also written to an in-memory cache and become immediately queryable.
+The cache is periodically written to disk in the form of [TSM](#time-structured-merge-tree-tsm) files.
+As TSM files accumulate, they are combined and compacted into higher level TSM files.
+
+Points can be sent individually; however, for efficiency, most applications send points in batches.
+Points in a POST body can be from an arbitrary number of series, measurements, and tag sets.
+Points in a batch do not have to be from the same measurement or tagset.
+
+## Write Ahead Log (WAL)
+
+The **Write Ahead Log** (WAL) ensures durability by retaining data when the storage engine restarts.
+It ensures that written data does not disappear in an unexpected failure.
+When a client sends a write request, the following steps occur:
+
+1. Append write request to the end of the WAL file.
+2. Write data to disk using `fsync()`.
+3. Update the in-memory cache.
+4. Return success to caller.
+
+`fsync()` takes the file and pushes pending writes all the way to the disk.
+As a system call, `fsync()` has a kernel context switch which is computationally expensive, but guarantees that data is safe on disk.
+
+{{% note%}}
+Once you receive a response to a write request, your data is on disk!
+{{% /note %}}
+
+When the storage engine restarts, the WAL file is read back into the in-memory database.
+InfluxDB then answers requests to the `/read` endpoint.
+
+## Cache
+
+The **cache** is an in-memory copy of data points currently stored in the WAL.
+Points are organized by key, which is the measurement, tag set, and unique field.
+Each field is stored in its own time-ordered range.
+Data is not compressed in the cache.
+The cache is recreated on restart by re-reading the WAL files on disk back into memory.
+The cache is queried at runtime and merged with the data stored in TSM files.
```

Comment:
The **cache** is an in-memory copy of all data points in the WAL. The cache:
- Organizes points by key (m, ts, and unique f) and fields are stored in time ordered range.
- Stores uncompressed data.
- Gets updates from the WAL file each time the storage engine restarts